### PR TITLE
Fix duplicate XLA compilation of jit(train_step)

### PR DESCRIPTION
**Pull request by olupton**
# Description

The AoT compilation used to extract memory stats previously caused a redundant compilation due to mismatched context managers. Skip the memory stats entirely when AutoPGLE is enabled as there is no nice way of getting the statistics up-front without causing redundant compilation.

# Tests

With https://docs.jax.dev/en/latest/config_options.html#jax_log_compiles this can be seen to reduce the number of training step compilations from 2 to 1 without AutoPGLE, or from 3 to 2 with AutoPGLE.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/trainers/pre_train/train.py [hunk 1]
```diff
@@ -958,7 +958,10 @@ def train_loop(config, recorder, state=None):
       params_shardings,
   )
 
-  with jax.set_mesh(mesh), mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+  # Do not enter the legacy `mesh` context manager here: the training loop calls
+  # p_train_step without it, and the mismatch in jit's tracing-cache key would
+  # cause train_step to be traced and compiled a second time on the first step.
+  with jax.set_mesh(mesh), nn_partitioning.axis_rules(config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/maxtext/trainers/pre_train/train.py:965]: @khatwanimohit is the diloco implementation still blocked by this?

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -972,7 +975,12 @@ def train_loop(config, recorder, state=None):
     else:
       lower_args = (state, shaped_batch)
     maxtext_utils.maybe_dump_jaxpr(config, p_train_step, lower_args)
-    if config.compiled_trainstep_file == "":  # compile only when there is no pre-compiled file loaded
+    if config.compiled_trainstep_file == "" and not jax.config.jax_enable_pgle:
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/trainers/pre_train/train.py:979]: I like this suggestion

File: src/maxtext/trainers/pre_train/train.py (hunk 1 above)
Comment (khatwanimohit) [reply src/maxtext/trainers/pre_train/train.py:965]: no, removing legacy mesh context is fine

**Review by NuojCheng — APPROVED** [review]

**Review by gobbleturk — APPROVED** [review]
